File: sagefs-vscode/src/codeLensProvider.ts
import * as vscode from "vscode";

/**
 * Provides CodeLens actions at ;; boundaries in F# files.
 * Shows "▶ Eval" above each code block.
 */
export class SageFsCodeLensProvider implements vscode.CodeLensProvider {
  private _onDidChangeCodeLenses = new vscode.EventEmitter<void>();
  readonly onDidChangeCodeLenses = this._onDidChangeCodeLenses.event;

  provideCodeLenses(
    document: vscode.TextDocument,
    _token: vscode.CancellationToken
  ): vscode.CodeLens[] {
    if (document.languageId !== "fsharp") {
      return [];
    }

    const lenses: vscode.CodeLens[] = [];
    let blockStart = 0;

    for (let i = 0; i < document.lineCount; i++) {
      const lineText = document.lineAt(i).text.trimEnd();
      if (lineText.endsWith(";;")) {
        const range = new vscode.Range(blockStart, 0, i, lineText.length);
        lenses.push(
          new vscode.CodeLens(range, {
            title: "▶ Eval in SageFs",
            command: "sagefs.evalRange",
            arguments: [range],
          })
        );
        blockStart = i + 1;
      }
    }

    return lenses;
  }

  refresh(): void {
    this._onDidChangeCodeLenses.fire();
  }
}

File: sagefs-vscode/src/CodeLensProvider.fs
module SageFs.Vscode.CodeLensProvider

open Fable.Core
open Fable.Core.JsInterop
open Vscode

module Blocks = SageFs.Vscode.CodeBlocks

/// Creates a CodeLens provider object compatible with VSCode's API.
/// Shows "▶ Eval" at the start of each code block — either ;; delimited or blank-line separated.
/// Respects density setting: disabled in Minimal and Normal modes.
let create () =
  createObj [
    "provideCodeLenses" ==> fun (doc: TextDocument) (_token: obj) ->
      let cfg = Workspace.getConfiguration "sagefs"
      let density = cfg.get("density", "full")
      match density with
      | "minimal" | "normal" -> [||]
      | _ ->
      let lenses = ResizeArray<CodeLens>()
      let blocks = Blocks.getAllBlockRanges doc
      for blockStart, _blockEnd in blocks do
        let range = newRange blockStart 0 blockStart 0
        let cmd = createObj [
          "title" ==> "▶ Eval"
          "command" ==> "sagefs.eval"
          "arguments" ==> [| box blockStart |]
        ]
        lenses.Add(newCodeLens range cmd)
      lenses.ToArray()
  ]
